### ISSUE TITLE
Tracking issue for channel selection

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#27800]
This is a tracking issue for the unstable `mpsc_select` feature in the standard library. Known points:
- The API is unsafe, and should probably never be stabilized.
- Servo apparently makes very heavy use of this API, so it can't be removed without replacement currently.
- The correctness of the implementation is somewhat dubious, it's incredibly complicated.
- The macro-based approach for selection is dubious.
- This is not a scalable implementation of selection (e.g. "poll everything each time wait() is called")
- This drags down some other channel-related code with a few mutexes in otherwise lock-free code.

**Comment by jdm** [comment]
cc me

**Comment by steveklabnik** [comment]
Is it worth investigating @BurntSushi 's https://github.com/BurntSushi/chan ?

**Comment by steveklabnik** [comment]
I also think @carllerche was supporting the idea of removing channels from the stdlib

**Comment by alexcrichton** [comment]
Note that I'm not advocating removing channels from the standard library (plus they're stable). Removing selection is also probably not an option as it's so closely tied to the implementation details of channels.

One of the major points of complexity of the current implementation is that it's all basically 99% lock free (giving us a nice performance boost in theory). I believe @BurntSushi's implementation (please correct me if I'm wrong!) is largely based on mutexes and condition variables.

**Comment by BurntSushi** [comment]
@alexcrichton Correct. I doubt very much that `chan` will ever be lock free because I'm not convinced the semantics I want can be done with them. It's also *much* easier to implement. :-)

**Comment by nagisa** [comment]
I personally prefer the epoll-like design as implemented in [comm](https://github.com/mahkoh/comm).

**Comment by carllerche** [comment]
Streams in [Eventual](github.com/carllerche/eventual) are lock free :)

Re: removing channels from stdlib, as @alexcrichton points out, they are stable. I would not worry about adding select to them. They are good as a starting point. If somebody requires more complex concurrency semantics, they can use a lib.

**Comment by tripped** [comment]
Just hopping on to mention #12902 as one of the current weaknesses of the macro approach to select!.

**Comment by softprops** [comment]
Fwiw, channel selection with golang channels is part of what makes golang concurrency story so alluring to newcomers to go. For a newcomer to a safe concurrent language like rust, needing to mull over which 3rd party concurrency crate is popular this week to pick for actually writing concurrent code feels onerous. Having primatives like channels and channel selection in the std lib make rust more attractive for concurrent applications than alternatives like go and make crates a little more interoperable than resolving conflicting implementation issues betwwen 3rd party options.

**Comment by BurntSushi** [comment]
> Fwiw, channel selection with golang channels is part of what makes golang concurrency story so alluring to newcomers to go.

@softprops This doesn't address your primary concerns, but FYI, [`chan_select!`](http://burntsushi.net/rustdoc/chan/#example:-select-on-multiple-channel-sends/receives) in the `chan` crate was built to specifically match the semantics of Go's select.

**Comment by softprops** [comment]
@burntsushi nice!

**Comment by jonhoo** [comment]
Correct me if I'm wrong, but doesn't the macro approach also have the drawback that selecting over a dynamic set of channels isn't easily expressible?

**Comment by szagi3891** [comment]
Best to get rid of the macro select.
I made a working prototype that shows that it is possible:
https://github.com/szagi3891/channels_async-rs/blob/master/examples/select.rs#L76
Without any section unsafe.

**Comment by canndrew** [comment]
Here's a wacky idea: add support for [generically sized tuples](https://github.com/rust-lang/rfcs/pull/1582) and [anonymous enums](https://github.com/rust-lang/rfcs/pull/1154). Then it will be possible to select across an arbitrary number of channels of different types without macros:

    let ch0: Receiver<i32> = ...;
    let ch1: Receiver<String> = ...;
    match select((ch0, ch1)) {
        (x|!) => println!("Got an i32: {}", x),
        (!|y) => println!("Got a String: {}", y),
    }

**Comment by BurntSushi** [comment]
Is there anyone still using this API? I don't see any obvious path to stabilization. It would be nice to just remove it.

**Comment by alexcrichton** [comment]
@jdm is this still being used in Servo?

**Comment by SimonSapin** [comment]
grep finds 4 instances of `select!` in servo master.

**Comment by mattgreen** [comment]
I'd like to remove this from watchexec, but I'm unsure what to move to.

I'm usually watching for either control-c or file system inputs at all times, since I have cleanup to do when the user hits control-c. `select!` might be unstable but I can at least accomplish this.

**Comment by softprops** [comment]
I've had only good experiences with https://github.com/BurntSushi/chan

**Comment by szagi3891** [comment]
@mattgreen :
I think you should be able to replace this selecta! macoro of one channel.

This channel type will be :
enum Message {
    CtrlC,
    NormalMessage(....),
}

The main thread would send in theright time message Message::CtrlC.
The remainder of the program would send messages Message::NormalMessage

It makes sense ?

**Comment by antrik** [comment]
Servo's `ipc-channel` library ( https://github.com/servo/ipc-channel ) also uses `mpsc_select` (for its `inprocess` back-end) -- which is a bit of a pain, since platforms relying on this back-end can't use `ipc-channel` with stable Rust. (See https://github.com/servo/ipc-channel/issues/118 )

It should be noted though that this back-end is not a real implementation, but rather just a partial workaround for platforms that don't have a "real" inter-process implementation yet. While it was originally created for Windows, most likely Windows will be getting a proper implementation soonish. ( https://github.com/servo/ipc-channel/pull/108 )

Right now the `inprocess` back-end is also being used for Android -- though I'm not sure why really...

So while this back-end might continue to be useful during bring-up of new platforms (depending on whether Servo keeps the option of running in a single process), it's not exactly a first-class citizen I'd say.

**Comment by antrik** [comment]
I'd like to point out that https://github.com/BurntSushi/chan -- which is being touted as a replacement for `mpsc` here -- doesn't really seem like a valid alternative to me: unlike `mpsc`, it has clonable and sharable receivers.

This might seem convenient for *some* use cases; but I'm not convinced it's a real win over just wrapping the receiver in a `Mutex<>` or `Arc<Mutex<>>` yourself when you *know* it's save and necessary for your application. (As mutexes are safe and easy to use in Rust -- unlike in Go -- I don't see any motivation for channels to double as a locking substitute...)

More importantly though, for the vast majority of use cases, sharing the receiver is actually *not* desirable: rather, it often indicates bugs -- and thus not rejecting this by default is indeed more error prone!

(Not even considering implementation overhead for something that's not needed at all in most cases...)

**Comment by BurntSushi** [comment]
@antrik Please see the README for `chan`. Notably:

> The purpose of this crate is to provide a safe concurrent abstraction for communicating sequential processes. Performance takes a second seat to semantics and ergonomics. If you're looking for high performance synchronization, chan is probably not the right fit.

There are a *number* of reasons why it doesn't necessarily replace `std::sync::mpsc`. Nevertheless, if you want to write *in a CSP style in Rust*, then I do think it's your only option on stable Rust because it gives you `chan_select!`. I suspect that's why people are bringing it up (which seems reasonable to me).

**Comment by antrik** [comment]
@BurntSushi as long as there is a general understanding that `chan` is a special-purpose library rather than a general replacement, that's perfectly fine -- I'm just concerned that there seems to be an air here of "we don't need to fix `mpsc` since there is `chan`", which I find rather disconcerting...

**Comment by antrik** [comment]
To be perfectly clear, my point is that I believe there is a genuine need for `mpsc` to have a stable select; or failing that, at the very least for some other library that can indeed serve as a real drop-in replacement. (And is prominently featured as such -- I for one still think that the prevalent "go look around for some random crate that seems like it might fit this very basic need" mindset is seriously hurting the Rust ecosystem...)

**Comment by BurntSushi** [comment]
I don't think anyone's really contesting that though? I think these are the facts on the ground as I know them:

* `mpsc_select` hasn't seen any attention towards stabilizing it. I don't think there's even a viable proposal in existence at all.
* It seems the only thing keeping `mpsc_select` around is Servo.
* *Some* have found success by switching to `chan`.

**Comment by p-kraszewski** [comment]
As a Rust newcomer from Erlang/Elixir world - I can live with MPSC.

My question is - do you have any comments on @szagi3891 approach for a static list of channels?  Instead of selecting on let's say 2 channels, just `.receive` on one, carrying enum that is `match`ed and dispatched later? Any do-s, don't-s, don't care-s?

**Comment by barr1969** [comment]
**Select on Dynamic Lists of Channels**

As someone whose been using CSP for decades, I feel compelled to submit a plea for the ability to select across dynamic lists of channels. This is such a useful thing to be able to do. It's particularly useful in large systems that have to deal with failure. That sounds like I'm effectively supporting @canndrew's suggestion. Lists that are fixed at compile time diminsh what can be achieved.

**Locks**

On the topic of striving for a lock free implementation, I'm not convinced there's much merit in that. 

If a receiver is selecting across one or more channels then it is implicitly content to wait for messages to arrive. So there's no real penality there from the receiver's point of view. 

However, if the receiver never ends up waiting, and this is seen as a problem, then my argument is that one's program's architecture is inadequate; the implication is that the receiver is a performance bottle neck. The real solution is that there needs to be additional receivers, a separate channel to each one, and a means (e.g. either round robin, push pull, etc - the parallels with [ZeroMQ's](http://zeromq.org/) patterns are _not_ coincidental) to know to which one the next message should be sent. 

If performance is an issue, the ideal architecture is one where there's enough parallel receivers to juuuuust ensure that the sender never has to block. If locks are used as part of the channel implementation, then the sender's end of a channel should then always be finding that the locks are available (because there's enough receivers to keep up with the required data flow rate).

**Comment by jonhoo** [comment]
As a separate suggestion, the eventual `select` interface should also have a `try_recv()` equivalent, as well as a `recv_timeout()`. I currently call `try_recv()` manually on each channel a couple of times before calling `wait()` to avoid incurring a thread sleep/wakup cycle, and being able to instead call `try_recv` directly on the `Select` would be much better.

**Comment by jonhoo** [comment]
Another use-case that's been bothering me lately is if I have multiple channels, some of which are likely to always be non-empty. The current `Select` implementation does not give any kind of guarantee of fairness, and if the first channel mentioned in the `Select` always has an element available, the other channels will never be read from.

**Comment by adrianbrink** [comment]
Is this an unstable feature that I can already use on nightly?

**Comment by steveklabnik** [comment]
yes https://doc.rust-lang.org/stable/std/macro.select.html

**Comment by bstrie** [comment]
Those following this thread will probably be interested in https://github.com/crossbeam-rs/rfcs/pull/22 .

**Comment by stjepang** [comment]
`crossbeam-channel` has been just released: https://docs.rs/crossbeam-channel

This crate aims to be a faster and more featureful alternative to `std::sync::mpsc`. I believe it should basically tick the boxes of all requests that have been brought up in this thread. I encourage everyone to give it a shot and let me know what they think about it. :)

-------------------

Regarding `std::sync::mpsc`, I must agree with @carllerche here:

> Re: removing channels from stdlib, as @alexcrichton points out, they are stable. I would not worry about adding select to them. They are good as a starting point. If somebody requires more complex concurrency semantics, they can use a lib. 

Implementing fast channels is difficult enough, and adding selection on top of them makes implementation crazy complicated. Perhaps the standard library should offer very simple channels (without selection) that work as a starting point, deferring advanced use to external crates?

Honestly, I believe a better fit for the standard library would be a new dead-simple MPMC channel that internally uses locks and conditional variables. It'd support blocking and non-blocking send/recv operations, that's all. No selection, and no advanced algorithms aimed at high performance.

My impression is that the present design of `std::sync::mpsc` can nowadays be seen as a historical artifact that reflects the needs of Servo.

For example, asynchronous channels are highly optimized and have complex implementation, yet the synchronous variant got little attention - it is comparatively simple and much slower. This is probably because Servo needs fast async channels but doesn't care about sync channels at all.

Furthermore, selection is imperfect in a number of ways. There doesn't seem to be a hopeful path towards stabilization of selection, but it has been kept alve (in a permanent unstable state) for years because Servo still depends on it.

**Comment by SimonSapin** [comment]
> It seems the only thing keeping mpsc_select around is Servo.

I’m hoping to move Servo to using crossbeam-channel. If all goes well and that lands, I’ll propose deprecating mpsc selection in std in order to find out if this quote above is accurate.

**Comment by droundy** [comment]
I'd like to add that I'd like to eventually see a nice channel implementation in the standard library.  They can be a part of APIs and once a good select is enabled it really helps for everyone to use the same kind of channels.

**Comment by ssokolow** [comment]
@droundy The interoperability convention that seems to be developing in the Rust ecosystem is to standardize a set of types and traits, publish them to crates.io alone, and then have each implementation depend on them.

See, for example, the [http](https://docs.rs/http/0.1.1/http/) crate, which provides shared definitions for types like `Request` and `Response`.

**Comment by WiSaGaN** [comment]
@ssokolow I agree with this, and I think this is compatible to have default implementation in standard library. Although comparing to `http`, I feel channel interface is more fundamental to programming in general (golang has it baked in the language), thus should be in standard library at least eventually.

**Comment by jonhoo** [comment]
I agree that the channel trait probably should, if we decide to go down this route. In a sense it's almost like `Read` and `Write`. The place where it gets tricky is if you want to support something like `select` across a common API. That I think will be really difficult, if not impossible. But basic channel behaviour I totally agree fits into that paradigm nicely.

**Comment by stacktracer** [comment]
As a random lurker, I hope I'm not speaking out of turn ... but to me select is _the_ magic sauce that makes channels compelling -- the one thing about channels you can't easily get in other paradigms.

Channels without select ... meh.

**Comment by ExpHP** [comment]
Just noting here that I didn't bother with `select!` in #48056 because it seemed to me like it had ergonomical issues that extend well above and beyond its lack of trailing comma support, and that this is an inappropriate problem to be solving with a public `macro_rules!` macro in libstd.  I'm glad to see that stabilization is already considered more or less off the table.

**Comment by BatmanAoD** [comment]
@alexcrichton Why does the top-level post say that the API "should probably never be stabilized"? Is there some theoretical reason why `select` can't be done in an acceptable way?

**Comment by SimonSapin** [comment]
@BatmanAoD I think what was meant is that the API should not be stabilized *in its current form*.

Designing a better API is possible, arguably https://docs.rs/crossbeam-channel/0.2.3/crossbeam_channel/macro.select.html has done so. Now a logical next question would be: should this or some other channel selection API be added to the standard library? I’d say that it should not, and it’s better to suggest users try `crossbeam-channel` (or some other library) instead, since libraries on crates.io that can evolve independently from the language.

`std::sync::mpsc` exists today mostly because it was inherited from once upon a time before Rust 1.0, when Cargo or the `Send` trait did not exist yet. When green threads and channels were considered a core part of Rust’s concurrency story and were both built into the language. If it were proposed for addition today it would probably not be accepted.

**Comment by gterzian** [comment]
As someone who is relatively new to Rust, concurrency, or channels, I would like to share my own experience using `Select` and the broader `std::sync::mpsc` API.

All I know of channels is what I have learned contributing to Servo for almost two years now. I can say the following with certainty: If it wasn't for `std::sync::mpsc`, and `Select`, after two years I would still be scratching my head about how the various components in Servo communicate with each other. 

With regards to the internal implementations of channels, my knowledge is limited, let's just say that I was recently surprised to read that there was a lock inside of them...

However, I can tell you that for a newcomer, having a concept like `std::sync::mpsc` at your disposal is great. 

Here are the reasons:
1. It has "fast, async" channels, which are simply much easier to get started with than any blocking flavor.
2. It's in the standard library, so you feel confident to use them and you don't need to shop around.
3. It isn't a "mpmc" or some other more complicated flavor of channel. To be convinced, just google "Go concurrency patterns". 

And as you learn more about how to use them, you come to realize that there is a lot you can do with an 'mpsc' channel, including stuff that you'd thought would require a dedicated flavor. 

For example 'spmc', who says it has to be a single value? What about having a set of `Sender`, one for each of your consumer? (Incidentally, it seems that such `spsc` use of `std::sync::mpsc` is also the most efficient way to send messages [these days](https://github.com/stjepang/rfcs-crossbeam/blob/channel/text/2017-11-09-channel.md#benchmarks)).

Regarding 'mpmc', I'm assuming you'd want to have some business logic around such communication, so why not have a custom 'broker' component and implement the communication with a bunch of `std::sync::mpsc`? In some ways, Servo's [constellation](https://github.com/servo/servo/blob/master/components/constellation/constellation.rs) could be seen as a giant 'mpmc' channel. 

Is `Select` currently really that bad, or missing features? To me, the Go 'more complicated' variant is basically just another way to deal with blocking channels.  Who needs to select over a `send`, when it can never block? On the other hand, `Select` is certainly useful when you need to receive a message from different components, each represented by a `Receiver`. It is playing a big part in making [logic like this](https://github.com/servo/servo/blob/de48f25d0b0b5ecaf0499030f8617a3d37b3fde2/components/script/script_thread.rs#L919) easier to understand.

And regarding `Sender` or even `Receiver` needing to be `Sync`, it appears to me that in fact almost any communication pattern with a `std::sync::mpsc` can be implemented without the need for those two to be `Sync`. I wrote a blog about it last week: https://medium.com/@polyglot_factotum/rust-concurrency-patterns-communicate-by-sharing-your-sender-11a496ce7791

Also, you need to appreciate the beauty of using a `Receiver` as an iterator, allowing a component to 'quit' simply when all it's senders have been dropped elsewhere. To see how hard that is to do with a 'mpmc' channels, just read this https://blog.golang.org/pipelines(here is my version: https://medium.com/@polyglot_factotum/rust-concurrency-patterns-natural-born-pipelines-4d599e7612fc)

Go has made channels popular, and people often say that Go is more "simple" than Rust. `std::sync::mpsc` is a clear example of a simple approach that is easy to work with, and that has been battle-tested in a large and complex Rust project like Servo.

Couldn't we move toward a stable `Select`, as opposed to the direction of deprecation? I'd do the work myself if I could actually wrap my head around all that manual bookkeeping(that I am so happy my own code using channels doesn't have to deal with). Let me know if I can help in some other way...

**Comment by BurntSushi** [comment]
@gterzian I found it hard to follow your comment, but have you see [`crossbeam_channel`](https://docs.rs/crossbeam-channel/0.2.3/crossbeam_channel/)? A lot of thought went into that API. It would be interesting to see whether Servo could migrate to it. If not, why not?

**Comment by SimonSapin** [comment]
Servo has an open PR to move to crossbeam-channel: https://github.com/servo/servo/pull/19515. It is currently blocked on a bug that needs tracking down (and causes tests to fail).

**Comment by gterzian** [comment]
@BurntSushi What was hard to follow about my comment? Or do you mean that it was just too long to read?

The TLDR of my comment is this: `std::sync::mpsc` is absolutely awesome, and `Select` plays a part in that. I am happy with "fast, async" channels and only selecting over receivers. I also think one can build flavors other than `mpsc` using `mpsc` as a building block, so I don't need other flavor of channels or a `Select` implementation that would support those. 

I think crossbeam looks interesting, and it's selling points are feature I don't actually feel are missing from `std::sync::mpsc`. I also think a crate isn't a replacement for a standard library, unless it becomes part of the standard library. 

I would be saddened if the standard library were to lose a full-featured channel implementation.

**Comment by gterzian** [comment]
@SimonSapin What was the initial impetus for that PR? I am myself not aware of any problems experienced with channels inside Servo(perhaps with the exception of `ipc-channel` blocking on sending(see https://github.com/servo/servo/issues/14704). I my experience, `Select` has been reliable and useful. Were there specific problems that haven't been documented so far?

**Comment by SimonSapin** [comment]
The fact that using `std::sync::mpsc::Select` requires `unsafe` is a very strongly negative point IMO. Having `.wait()` return an integer that you have to compare to other things yourself also feels error-prone.

My personal motivation for https://github.com/servo/servo/pull/19515 is reducing the number of unstable features used in Servo, especially those that apparently do not have a path to stabilization.

**Comment by gterzian** [comment]
@SimonSapin I see, those are worthy goals in deed. 

So essentially, if `Select` were to ditch the unsafe `add`, and improve the `wait` api, it could be moved into stable? Would there be interest from someone to review this kind of work, a path towards merging it? I have the summer ahead of me, if someone is willing to get some pretty naive implementations thrown at them...

What is the story behind there not being a path towards stabilization? It certainly doesn't seem to be due to a lack of brains in the community writing channel implementations...

**Comment by BurntSushi** [comment]
@gterzian My suspicion is that folks will probably want to coalesce around `crossbeam-channel`. I'd say your first task would be convincing folks that `crossbeam-channel` is somehow insufficient or worse than a different implementation. I'd probably start by prototyping your desired API outside of std.

> What is the story behind there not being a path towards stabilization?

Because folks have been concentrating their efforts on experimenting and standardizing on APIs outside of `std` since there is no compelling reason (as of yet) for these APIs to actually be in the standard library. (You might disagree with what qualifies as a "compelling reason," but I'm not trying to be prescriptive or opinionated here. I'm more or less trying to be descriptive based on what I perceive the consensus to be. Others may feel free to chime in on that point though!)

**Comment by SimonSapin** [comment]
The Stable Rust ecosystem now has a channel library that supports channel selection, with in my opinion a superior API. Why is `std::sync::mpsc` worth spending effort on? Why is it important to be in the standard library? There are many other crates that are important to the ecosystem (for some idea of "important") but yet are not part of the standard library.

**Comment by ssokolow** [comment]
My main concern is that having `std::sync::mpsc` in `std` to satisfy API stability requirements but without a corresponding stable select API `std` will continue to cause misconceptions like RMS's.

**Comment by stjepang** [comment]
@gterzian

> It isn't a `mpmc` or some other more complicated flavor of channel. To be convinced, just google "Go concurrency patterns".

Perhaps this is not needed in Servo, but a lot of people have been requesting support for multiple consumers. They typically just do `Mutex<Receiver<T>>`, which is unergonomic and slow.

> Using a set of spsc channels, you can build your own spmc one(Incidentally, it seems that such spsc use of std::sync::mpsc is also the most efficient way to send messages these days).

Simulating a SPMC channel using SPSC channels is not easy (unless if you meant "broadcast" channels, akin to the [`bus`](https://docs.rs/bus) crate). Which consumer should a message be sent to? Maybe a random one, but what if some consumers consume messages faster than others?

Also, take a look at the updated benchmark results [here](https://github.com/crossbeam-rs/crossbeam-channel/tree/master/benchmarks#results).

> Is Select currently really that bad, or missing features?

`crossbeam-channel`'s select macro has a safe interface, the macro produces nice error massages, can receive from a dynamic set of receivers, and is faster than `mpsc::Select`.

> Who needs to select over a send, when it can never block?

Send can block in bounded channels (`mpsc::sync_channel`), so this is a nice to have feature.

> It is playing a big part in making logic like this easier to understand.

Consider [this](https://github.com/servo/servo/blob/de48f25d0b0b5ecaf0499030f8617a3d37b3fde2/components/script/script_thread.rs#L943) part in Servo where a message is selected from five channels. The equivalent code using `crossbeam-channel` would be:

```rust
let mut event = select! {
    recv(self.port, msg) => FromScript(msg.unwrap()),
    recv(self.control_port, msg) => FromConstellation(msg.unwrap()),
    recv(self.timer_event_port, msg) => FromScheduler(msg.unwrap()),
    recv(self.image_cache_port, msg) => FromImageCache(msg.unwrap()),
    recv(self.devtools_chan.map(|_| &self.devtools_port), msg) => FromDevTools(msg.unwrap()),
};
```

This is arguably much nicer and contains no unsafe code. :)

> And regarding Sender or even Receiver needing to be Sync, it appears to me that in fact almost any communication pattern with a std::sync::mpsc can be implemented without the need for those two to be Sync

Unfortunately, sometimes it's [not so easy](
https://github.com/rust-lang/rfcs/pull/1299#issuecomment-146361678).

> For example using a Receiver as an iterator, allowing a component to 'quit' simply when all it's senders have been dropped elsewhere. To see how hard that is to do with a mpmc channel, just read this blog.golang.org/pipelines (here is my version ).

Yes, that's very nice and `crossbeam-channel` can do this too. Probably the only reason why Go doesn't close channels automatically is because it lacks RAII. It doesn't have anything to do with channels being MPMC.

> Couldn't we move toward a stable Select, as opposed to the direction of deprecation? I guess there are perhaps some problems with the implementation, and I'm basically arguing in long form that the API is just fine.

That's been the plan for the past several years, but basically no progress has been made so far. And the reason is that it'd be very difficult to tweak the current implementation to expose a satisfying `Select` interface (ideally without `unsafe`).

> I my experience, Select has been reliable and useful. Were there specific problems that haven't been documented so far?

Tricky bugs are occasionally reported in `std::sync::mpsc` - [here's one](https://github.com/rust-lang/rust/issues/39364) currently open issue. These kinds of bugs are typically very difficult to track down. One commenter said:

> (And I can't help but note that it is not very confidence inspiring safety-wise for would-be newcomers to find crashes in the stdlib unfixed for a long time...)

Now, `crossbeam-channel` hasn't been battle-tested as much, but it's a clean slate implementation, has fewer `unsafe` blocks, and goes to great length to ensure correctness. We currently have [these tests](https://github.com/crossbeam-rs/crossbeam-channel/tree/master/tests):

* An extensive test suite for each channel flavor (bounded, unbounded, zero-capacity) and `select!`.
* A copy of all [tests from `std::sync::mpsc`](https://github.com/crossbeam-rs/crossbeam-channel/blob/master/tests/mpsc.rs).
* We're gradually porting all the [tests for Go channels](https://github.com/crossbeam-rs/crossbeam-channel/blob/master/tests/golang.rs) into Rust.
* Each test is [wrapped](https://github.com/crossbeam-rs/crossbeam-channel/blob/d31e47d104619b4e58a032bd10ddd0bedec7af16/tests/golang.rs#L888) into 5 slightly different flavors that attempt to trigger as many internal code paths as possible.

This currently amounts to 1200 tests total, and we're writing more.

> So essentially, if Select were to ditch the unsafe add, and improve the wait api, it could be moved into stable?

Yes, probably. Care will have be taken not to hurt performance when removing unsafe functions.

> What is the story behind there not being a path towards stabilization? It certainly doesn't seem to be due to a lack of brains in the community writing channel implementations...

It's just really difficult to design a nice, safe interface and make channels fast at the same time. It might be even more difficult now than is necessary due to certain (historical) design decisions in the internal implementation of `std::sync::mpsc`.

In order to stabilize `Select`, I think we'd have to put so much work that it simply wouldn't be justified given the fact we have crates with alternatives.

**Comment by ExpHP** [comment]
Beyond any issues with `Select` (which I wouldn't know about), there are also issues with the macro.

It looks like `match` syntax, but:

* No trailing commas. (easily fixed, but I didn't bother as it looked perma-unstable)
* It requires a comma even after braced match arms: (this requires an incremental muncher to fix)
  ```rust
  select! {
      a = rx1.recv() => {
          println!("{}", a.unwrap());
      },                              // <-- comma required
      a = rx2.recv() => panic!()
  }
  ```
* You get a weird type error if you accidentally write the same `rx` twice.  (This is because it rebinds the identifier to something else)

But moreover there is the fundamental issue that this is *unusually elaborate syntax* for a `std` macro.  It feels like something that should be in another crate.

**Comment by RalfJung** [comment]
> My main concern is that having std::sync::mpsc in std to satisfy API stability requirements but without a corresponding stable select API in std will continue to cause misconceptions like RMS's.

The best solution to this problem is deprecation, isn't it?

**Comment by BatmanAoD** [comment]
@SimonSapin That makes more sense -- I was confused by the use of the word "never", and this issue was linked in a post I was reading (by an author who apparently had the same misunderstanding) as support for a claim that there probably wouldn't ever be a stable version of Select:

 >     Note: the ‘Select’ feature is not stable, perhaps never will be, see https://github.com/rust-lang/rust/issues/27800.

https://medium.com/@polyglot_factotum/rust-concurrency-patterns-communicate-by-sharing-your-sender-11a496ce7791

**Comment by ssokolow** [comment]
@RalfJung I'm uncertain.

If I were to use something with multiple parties on either end, I'd probably use `crossbeam-channel` so I didn't have to hack something up for the SPMC direction but it'd still be nice to have a high-performance SPSC channel in `std`, especially when it's currently the [fastest](https://github.com/stjepang/rfcs-crossbeam/blob/channel/text/2017-11-09-channel.md#benchmarks) option in that use case.

Perhaps just deprecating the multi-consumer aspect of it and offering an un-deprecated version under a new name?

**Comment by stjepang** [comment]
@ssokolow

> so I didn't have to hack something up for the SPMC direction but it'd still be nice to have a high-performance SPSC channel in std, especially when it's currently the fastest option in that use case.

I've managed to close this performance gap in `crossbeam-channel` - see the [newest benchmarks](https://github.com/crossbeam-rs/crossbeam-channel/tree/master/benchmarks#results).

**Comment by SimonSapin** [comment]
@BatmanAoD Well, it is *also* a possible future that `Select` will be deprecated, eventually removed, and never replaced in the standard library.

@ssokolow Re fastest, I think that’s with an older version of crossbeam-channel. @stjepang has linked above to updated benchmarks here: https://github.com/crossbeam-rs/crossbeam-channel/tree/master/benchmarks#results

**Comment by gterzian** [comment]
@stjepang Thank you, finally someone who like to write as much as I do.

> Simulating a SPMC channel using SPSC channels is not easy (unless if you meant "broadcast" channels, akin to the bus crate). Which consumer should a message be sent to? Maybe a random one, but what if some consumers consume messages faster than others?

I'm less concerned about underlying implementation details, vs conceptual behavior. So to me, having a set of SPSC channels simulate an SPMC doesn't need to involve any sort of underlying implementation guarantees. I don't need a 'fair' SPMC or something like that, just looping over the `Sender` and sending them work will do. 

Surely that is conceptually better than sharing Mutex<Receiver<T>> to each worker?

If I need such a SPMC queue offering these guarantees, I will indeed look for a specialized crate. 

----

With regards to the issue with [hyper](https://github.com/rust-lang/rfcs/pull/1299#issuecomment-146361678), that is a result of the design of `hyper`, and [here](https://github.com/stjepang/rfcs-crossbeam/blob/channel/text/2017-11-09-channel.md) you've also highlighted this occurrence  in Servo with the [WorkletExecutor](https://github.com/servo/servo/blob/38f4ae80c4b456b89ee33543c8c6699501696c9c/components/script/dom/paintworkletglobalscope.rs#L234). 

The problem here is more subtle, but it goes to the heart of how channels are used and abused.

The idea of channels is that you're not going to share them via references, you're going to move the `Sender`, or the clone to dedicated threads. That matches the idea of 'one sender per component, probably in their own thread'. it's very much `Send`(`thread::spawn` only requires `Send`).

Both examples of needing `Sync` involved some sort of "handler object" with references to it shared across threads(I can't tell what `hyper` does exactly, but in Servo I can see the `Arc<>`). 

If you're going to put something in an `Arc<>`, can you really complain about having to put locks around stuff inside of it? 

The solution in the hyper case, could be for hyper to provide their own channel to be used inside handlers. I mean, they're offering you a kind of 'runtime' with those handlers aren't they? So if messaging inside of those makes sense, they could supply such a solution. 

----

On iterating on `Reiceiver`, and 'quitting' when senders are dropped,  thank you for pointing out this is possible in `mpmc` and other scenarios too. 

> This currently amounts to 1200 tests total, and we're writing more.
Much appreciated. 

> That's been the plan for the past several years, but basically no progress has been made so far. And the reason is that it'd be very difficult to tweak the current implementation to expose a satisfying Select interface (ideally without unsafe).

Ok, thanks for the info.

> Tricky bugs are occasionally reported in std::sync::mpsc - here's one currently open issue. 

I will check it out, thanks. 

> (On the select macro of crossbeam) This is arguably much nicer and contains no unsafe code. :)

Yes I agree, it certainly looks nicer. 

> Send can block in bounded channels (mpsc::sync_channel), so this is a nice to have feature.

Yes I understand, and what I meant is rather that just having a select over a set of receivers is "good enough", especially assuming the unbounded variant sees more usage.

**Comment by ssokolow** [comment]
@stjepang Nice! Thanks for that. :)

**Comment by stjepang** [comment]
@gterzian

You're implying that `std::sync::mpsc` was intentionally designed the way it was because we really wanted that particular design. That's just not true.

`std::sync::mpsc` is a MPSC channel primarily because MPMC would require locks (that'd make it slow) or deferred memory reclamation (difficult problem - a viable solution appeared only a few years ago).

The reason why `Sender` doesn't implement `Sync` yet is because it's difficult (mainly because unbounded channels switch between three internal flavors: oneshot, spsc, and shared). We really want to make it implement `Sync`, just don't know how.

Interestingly, `SyncSender` does implement `Sync`, and that's only because it uses locks behind the scenes so that made it easy. Locks also make the channel very slow, and the reason why we use locks is because nobody has put in the effort of improving the implementation yet.

A lot of design choices behind `std::sync::mpsc` were the result of unfortunate technical difficulties so we had to make tradeoffs. The channels are not optimized as well as they could be and they don't fit all the use cases we want them to. See the discussion [here](https://github.com/rust-lang/rust/pull/42397#issuecomment-315812623) for more information.

> Besides this bias, a standard library results in 'shared vocabulary' across the language, which for the most basic building blocks is a good thing, even for a systems language like Rust.

I agree with this, but...

We already have an established 'shared vocabulary' across the ecosystem and that seems to be working very well: `serde` is for serialization, `rayon` is for data parallelism, `tokio` is for async IO, `http` is for HTTP, `crossbeam` is for concurrency, `rand` is for random number generation, etc.

Adding dependencies to Rust programs is easy. As soon as we get Rust 2018, adding a dependency will become as simple as `cargo add crossbeam`. That means there's not as strong motivation for moving stuff into the standard library as in other languages. The aforementioned crates are not exactly second-class citizens to Rust - they've been widely accepted as the standard set of basic tools.

**Comment by gterzian** [comment]
@stjepang Thanks, it's very interesting to read more about the background of `std::sync::mpsc`. 

You're right that I don't actually know anything about the original thinking behind the design and the implementation travails of `std::sync::mpsc`, so I should rather explicitly say that I'm trying to document what I see as a form of customary usage, mainly from observations in Servo, which might have grown organically over time and not reflect the initial design considerations for channels(which is actually very interesting).

This usage, as I see it, requires `Send` channels by default such as in

1. the set of 'webpage event-loops' managed by Servo is represented by a `HashMap<TopLevelBrowsingContextId, HashMap<Host, Weak<EventLoop>>>`, where `EventLoop` is actually a wrapper around a `IpcSender<ConstellationControlMsg>`.
2. The [`EmbedderProxy`](https://github.com/servo/servo/blob/master/components/embedder_traits/lib.rs#L35), again a wrapper around a `Sender`. 
3. In general a lot of the attributes on the [`constellation`](https://github.com/servo/servo/blob/78e90d6358d2ce2e06862dec9eae1e72bbe2bfab/components/constellation/constellation.rs#L170), which is like a custome `mpmc` channels really, are a lot of `Sender` and `Receiver`(with ipc variants). 

The [`ScriptThread`](https://github.com/servo/servo/blob/78e90d6358d2ce2e06862dec9eae1e72bbe2bfab/components/script/script_thread.rs#L400) has a few `Sender` wrapped inside a `Box`, I think because those that are passed around other threads to schedule 'tasks' back on the task queue of a script thread.

And incidentally we can even find some `Arc<Mutex<IpcSender>>` like in the [`CSSErrorReporter`](https://github.com/servo/servo/blob/12d82fe621947c2b29b12d9c9e9a81b0f4baa117/components/script_layout_interface/reporter.rs#L21)

The whole gamut is really there, and I'd say it doesn't feel weird to see a `Sender` wrapped inside a `Box` directly or via a `Arc<Mutex<SomethingContainingASender>>` for specific use cases. 

In the [`main` function](https://github.com/servo/servo/blob/78e90d6358d2ce2e06862dec9eae1e72bbe2bfab/components/servo/lib.rs#L131) where all the various components are created, you can see a lot of channels being created and moved around, without any being wrapped in an `Arc`.

----

> We already have an established 'shared vocabulary' across the ecosystem and that seems to be working very well

Thanks for sharing that, I think I can now appreciate better that the standard library in Rust isn't as important as in some other languages, like Python, and that even if `std::sync::mpsc` were to be entirely deprecated, Rust would still have vibrant 'shared vocabulary' covering many different approaches to concurrency.

I look forward to using `crossbeam` in Servo when that PR merges, and getting to know some of those different flavors of channels. Maybe there is something I can do to help fix those last remaining bugs...

**Comment by SimonSapin** [comment]
With https://github.com/servo/servo/pull/21325 merged (thanks @stjepang and @gterzian!), Servo does not use this feature anymore.

Do we know of other projects that do? Would it be useful to do a crater run (with something like renaming the feature flag) to find out?

@rust-lang/libs How do you feel we should proceed with this feature?

**Comment by alexcrichton** [comment]
FWIW https://github.com/rust-lang/rust/pull/54228 seems like a good way to proceed. I think if that's small enough we can probably just outright delete the feature, and if it's large we can do a round of messaging with a schedule to delete.

**Comment by SimonSapin** [comment]
I’m in favor of deleting eventually, but since that part of the code isn’t being actively worked on there isn’t much to gain by deleting sooner rather than later. We can keep it with a deprecation warning for a few release cycles, just in case there are users not covered by Crater to leave them some more time to migrate. (This migration took a while to land in Servo.)

I my opinion #54228 can help figure out if we want to keep delaying even the deprecation warning like we did so far.

**Comment by SimonSapin** [comment]
Crater has found three regressions. One spurious, one that I submitted a PR to fix, and one in a benchmark measuring specifically the performance of the `select!` macro: https://github.com/rust-lang/rust/pull/54228#issuecomment-421759312

So let’s formally propose:

* Emit a deprecation warning for usage of this feature
* Wait for a few release cycles for remaining users to have an opportunity to see the warning and potentially comment here
* Then, unless new information or arguments have come up, remove the feature.
* According to this issue’s original message, this would allow removing some mutexes in the rest of the channels code to make it lock-free. Would someone be interested in working on that?

@rfcbot fcp close

**Comment by SimonSapin** [comment]
Next is landing a deprecation warning.

**Comment by SimonSapin** [comment]
Triage: this unstable feature was deprecated in 1.32, so it is fine to remove any time.

**Comment by stjepang** [comment]
I published a blog post titled *[Proposal: New channels for Rust's standard library](https://stjepang.github.io/2019/03/02/new-channels.html)* discussing what to do with channels in the standard library.

How should we proceed about resolving this issue?

Bug #39364 is still present, which @stepancheg attempted to fix in PR #42883 but got stuck waiting on removal of `mpsc_select`. Perhaps now is a good time to remove `mpsc_select` and revive that PR.

Another option is to switch the whole implementation to https://github.com/stjepang/new-channel. As a bonus, we get performance improvements, `impl Sync for Sender`, `SyncSender::send_timeout`, more tests, lower compilation time, and fewer unsafe blocks.

Some people have suggested we just deprecate the whole `mpsc` module and leave channels out of the standard library.

Finally, the last option is to write a RFC for `std::sync::channel` that would eventually replace `std::sync::mpsc`. This would also allow us to fix #42883 and bring all the other improvements to `std::sync::mpsc` in the meantime.

**Comment by SimonSapin** [comment]
Oh I didn’t realize there was already work blocked on the removal of selection. If you want to send a removal PR feel free to ping me for review. We’re already good to go process-wise.

Replacing the internals of `std::sync::mpsc` and/or deprecating it is also interesting, but not relevant to this issue. It should probably be discussed on internals.rlo and then turned into an RFC.

**Comment by stjepang** [comment]
@stepancheg Would you be interested in giving PR #42883 another shot and removing selection along the way?

For followers of this thread, here's the the last [status report](https://github.com/rust-lang/rust/issues/39364#issuecomment-409622394) on the bug fix.

**Comment by theronic** [comment]
What's the succession story for waiting on two channels and taking from the first one that yields a message?

**Comment by SimonSapin** [comment]
Consider using another library such as https://crates.io/crates/crossbeam-channel

**Comment by havelund** [comment]
Sure, but still: being able to select which channel to read from seems to be a fundamental feature that should follow a message passing paradigm based on channels. I tried using mutexes and that was a disaster since they are not reentrant.

**Comment by BurntSushi** [comment]
@havelund As linked above, use a library: https://crates.io/crates/crossbeam-channel

**Comment by havelund** [comment]
Yes thanks. I found it. It seems to work fine.

**Comment by sdegutis** [comment]
For what it's worth, I ended up at this issue because I was trying to find out what happened to the `select!` macro in `std::sync::mpsc` which would be useful to me right about now. And only through this issue did I find out about `crossbeam-channel` which I'm switching to for its `select!`. Would be nice to have something like that in the standard library.

**Comment by dralley** [comment]
Is there still any discussion about deprecating ```std::sync::mpsc``` in favor of something based on ```crossbeam-channel```?  I haven't seen any discussion since just after the blog post came out.

**Comment by SimonSapin** [comment]
`std::sync::mpsc` (or at least its public API) is here to stay since it’s `#[stable]`. At most we could make its usage emit deprecation warnings, but what positive outcome would that achieve? That would be to balance against ecosystem churn.

**Comment by jonhoo** [comment]
@SimonSapin I read @dralley's proposal as replacing the internals, not the API, but could be wrong?

**Comment by dralley** [comment]
Not my proposal just to be clear, I'm just curious as to whether any action items had ever come out of that discussion.

**Comment by Ace-Cassidy** [comment]
As someone who came to love Go, coded in it professionally for almost a year, and subsequently fell out of love with it (generics, performance, package management, code gen hell), the lack of Select in the standard library in Rust is what is preventing me from adopting it.

If I need high performance and I'm going to have to rely on external code for concurrency support anyways, I might as well go with the path of least resistance and write in C.

**Comment by BurntSushi** [comment]
@Ace-Cassidy You just need to add `crossbeam-channel = "0.4"` to your `Cargo.toml` and you should be good to go.

**Comment by ssokolow** [comment]
@Ace-Cassidy Rust is designed to have a minimal standard library so that different components can evolve their APIs and bump their versions independently from each other.

You'll also need external crates (maintained by Rust team members) for things like random number generation ([`rand`](https://lib.rs/crates/rand)), regular expression support ([`regex`](https://lib.rs/crates/regex)), or pre-built bindings for the parts of the POSIX API that don't have equivalents on Win32 ([`libc`](https://lib.rs/crates/libc)).
